RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc

```diff
@@ -0,0 +1,1038 @@
+ */ 
+
+#include "CSCSegAlgoRU.h" 
+#include "CSCSegFit.h"
+#include "Geometry/CSCGeometry/interface/CSCLayer.h" 
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h" 
+#include "FWCore/ParameterSet/interface/ParameterSet.h" 
+#include "FWCore/MessageLogger/interface/MessageLogger.h" 
+
+#include <algorithm> 
+#include <cmath> 
+#include <iostream> 
+#include <string> 
+
+CSCSegAlgoRU::CSCSegAlgoRU(const edm::ParameterSet& ps)
+  : CSCSegmentAlgorithm(ps), myName("CSCSegAlgoRU"), sfit_(0) { 
+
+  doCollisions = ps.getParameter<bool>("doCollisions"); 
+  chi2_str_   = ps.getParameter<double>("chi2_str_"); 
+  chi2Norm_2D_   = ps.getParameter<double>("chi2Norm_2D_");    
+  dRMax       = ps.getParameter<double>("dRMax"); 
+  dPhiMax        = ps.getParameter<double>("dPhiMax"); 
+  dRIntMax   = ps.getParameter<double>("dRIntMax"); 
+  dPhiIntMax    = ps.getParameter<double>("dPhiIntMax"); 
+  chi2Max        = ps.getParameter<double>("chi2Max"); 
+  wideSeg        = ps.getParameter<double>("wideSeg"); 
+  minLayersApart = ps.getParameter<int>("minLayersApart"); 
+
+  LogDebug("CSC") << myName << " has algorithm cuts set to: \n" 
+		  << "--------------------------------------------------------------------\n" 
+		  << "dRMax     = " << dRMax << '\n' 
+		  << "dPhiMax      = " << dPhiMax << '\n' 
+		  << "dRIntMax = " << dRIntMax << '\n' 
+		  << "dPhiIntMax  = " << dPhiIntMax << '\n' 
+		  << "chi2Max      = " << chi2Max << '\n' 
+		  << "wideSeg      = " << wideSeg << '\n' 
+		  << "minLayersApart = " << minLayersApart << std::endl; 
+
+  //reset the thresholds for non-collision data 
+  if(!doCollisions){ 
+    dRMax = 2.0; 
+    dPhiMax = 2*dPhiMax; 
+    dRIntMax = 2*dRIntMax; 
+    dPhiIntMax = 2*dPhiIntMax; 
+    chi2Norm_2D_ = 5*chi2Norm_2D_; 
+    chi2_str_ = 100; 
+    chi2Max = 2*chi2Max; 
+  } 
+} 
+
+std::vector<CSCSegment> CSCSegAlgoRU::run(const CSCChamber* aChamber, const ChamberHitContainer& rechits){ 
+  theChamber = aChamber;  
+  return buildSegments(rechits);  
+} 
+
+std::vector<CSCSegment> CSCSegAlgoRU::buildSegments(const ChamberHitContainer& urechits) { 
+  ChamberHitContainer rechits = urechits; 
+  LayerIndex layerIndex(rechits.size()); 
+  int recHits_per_layer[6] = {0,0,0,0,0,0}; 
+
+  //skip events with high multiplicity of hits 
+  if (rechits.size()>150){ 
+    return std::vector<CSCSegment>(); 
+  } 
+
+  int iadd = 0; 
+  for(unsigned int i = 0; i < rechits.size(); i++) { 
+    recHits_per_layer[rechits[i]->cscDetId().layer()-1]++;//count rh per chamber 
+    layerIndex[i] = rechits[i]->cscDetId().layer(); 
+  } 
+
+  double z1 = theChamber->layer(1)->position().z(); 
+  double z6 = theChamber->layer(6)->position().z(); 
+
+  if ( z1 > 0. ) { 
+    if ( z1 > z6 ) {  
+      reverse(layerIndex.begin(), layerIndex.end()); 
+      reverse(rechits.begin(), rechits.end()); 
+    }     
+  } 
+
+  else if ( z1 < 0. ) { 
+    if ( z1 < z6 ) { 

```

Comment:
You are right. This part was inherited from the SK algo and I didn't pay attention to this. I can change this too.